docs/dotnet/dotnet-http-metrics.md

```diff
@@ -0,0 +1,257 @@
+| [`server.socket.address`](../general/attributes.md) | string | Physical server IP address or Unix socket address. If set from the client, should simply use the socket's peer address, and not attempt to find any actual server IP (i.e., if set from client, this may represent some proxy server instead of the logical server). | `10.5.3.2` | Recommended: If different than `server.address`. |
+| [`url.scheme`](../url/url.md) | string | The [URI scheme](https://www.rfc-editor.org/rfc/rfc3986#section-3.1) component identifying the used protocol. | `https`; `ftp`; `telnet` | Recommended |
+
+**[1]:** HTTP request method value SHOULD be "known" to the instrumentation.
+By default, this convention defines "known" methods as the ones listed in [RFC9110](https://www.rfc-editor.org/rfc/rfc9110.html#name-methods)
+and the PATCH method defined in [RFC5789](https://www.rfc-editor.org/rfc/rfc5789.html).
+
+If the HTTP request method is not known to instrumentation, it MUST set the `http.request.method` attribute to `_OTHER` and, except if reporting a metric, MUST
+set the exact method received in the request line as value of the `http.request.method_original` attribute.
+
+If the HTTP instrumentation could end up converting valid HTTP request methods to `_OTHER`, then it MUST provide a way to override
+the list of known HTTP methods. If this override is done via environment variable, then the environment variable MUST be named
+OTEL_INSTRUMENTATION_HTTP_KNOWN_METHODS and support a comma-separated list of case-sensitive known HTTP methods
+(this list MUST be a full override of the default known method, it is not a list of known methods in addition to the defaults).
+
+HTTP method names are case-sensitive and `http.request.method` attribute value MUST match a known HTTP method name exactly.
+Instrumentations for specific web frameworks that consider HTTP methods to be case insensitive, SHOULD populate a canonical equivalent.
+Tracing instrumentations that do so, MUST also set `http.request.method_original` to the original value.
+
+**[2]:** `network.protocol.version` refers to the version of the protocol used and might be different from the protocol client's version. If the HTTP client used has a version of `0.27.2`, but sends HTTP version `1.1`, this attribute should be set to `1.1`.
+
+`http.request.method` has the following list of well-known values. If one of them applies, then the respective value MUST be used, otherwise a custom value MAY be used.
+
+| Value  | Description |
+|---|---|
+| `CONNECT` | CONNECT method. |
+| `DELETE` | DELETE method. |
+| `GET` | GET method. |
+| `HEAD` | HEAD method. |
+| `OPTIONS` | OPTIONS method. |
+| `PATCH` | PATCH method. |
+| `POST` | POST method. |
+| `PUT` | PUT method. |
+| `TRACE` | TRACE method. |
+| `_OTHER` | Any HTTP method that the instrumentation has no prior knowledge of. |
+<!-- endsemconv -->
+
+### Metric: `http.client.failed_requests`
```

Comment:
If I was to diagnose errors, I would look for a metric that represents an error counter tagged with the error reason. If we think it's easy for users to derive that counter I'm fine going with (A).

From a previous comment:

> I believe it should be easy, we can try with prometheus. Users would write a query like sum(http_request_duration) by http_response_status_code, error_reason

When designing instruments do we consider Prometheus to be the primary analysis tool, or are there other use cases of metrics consumption we should take into account?

Reply:
> If I was to diagnose errors, I would look for a metric that represents an error counter tagged with the error reason. If we think it's easy for users to derive that counter I'm fine going with (A).

The counter is already available along with the duration histogram. We still want  duration for failed requests, so essentially adding another counter for failed requests would mean duplicating a subset of the time series.

Duplication without clear value is not great.

> When designing instruments do we consider Prometheus to be the primary analysis tool, or are there other use cases of metrics consumption we should take into account?

We're making sure data is available and other backends can transform it. We can write a similar query with azure monitor, something like this:

```kql
customerMetrics
| where name == "http_request_duration"
| summarize avg(value) by customDimensions["http_response_status_code"], customDimensions["error_reason"], bin(timestamp, 1m)
| render columnchart 
```

this might need a bit of configuration to be shown properly, but essen
Any

Reply:
I see it as a tradeoff of discoverability vs. collected data size. If there is counter named "failed_stuff" that is much easier for a developer to spot in a list of metrics. I'm fine saying our starting point is to favor minimal data size at the expense of discoverability, but I think we'll want to take steps to mitigate it via docs, examples, pre-made dashboards, APM tooling, etc. Then we can look at feedback to see if developers finding it successfully or we need to take a stronger measure in the future.

In short, lets do (A) and then keep an eye out going forward if we've done enough to facilitate discoverability.

Reply:
resolving for now, but feel free to unresolved as you see fit